hspec-discover/src/Config.hs

```diff
@@ -32,14 +35,17 @@ usage prog = "\nUsage: " ++ prog ++ " SRC CUR DST [--module-name=NAME]\n"
 parseConfig :: String -> [String] -> Either String Config
 parseConfig prog args = case getOpt Permute options args of
     (opts, [], []) -> let
-        c = (foldl (flip id) defaultConfig opts)
-      in
-        if (configNoMain c && isJust (configFormatter c))
-           then
-             formatError "option `--formatter=<fmt>' does not make sense with `--no-main'\n"
-           else
-             Right c
+        c = foldl (flip id) defaultConfig opts
+      in validateConfig c
     (_, _, err:_)  -> formatError err
     (_, arg:_, _)  -> formatError ("unexpected argument `" ++ arg ++ "'\n")
   where
     formatError err = Left (prog ++ ": " ++ err ++ usage prog)
+
+    validateConfig c | configHelp c = Left (helpMessage prog)
```

Comment:
Help messages should go to `stdout` and the program should exit with exit code `0`.  Returning a `Left` here will result in help messages being treated the same way as errors.  They will be printed to `stderr` and the program will terminate with `exitFailure`.